lib/src/io/controller.c

```diff
@@ -0,0 +1,109 @@
+/* The comment below is needed for this file to be picked up by generate_ld */
+/* RAM_POS: 0x800CCC20 */
+
+#include "PRinternal/macros.h"
+#include "PR/os_internal.h"
+#include "PRinternal/controller.h"
+#include "PRinternal/siint.h"
+
+#define HALF_MIL_CYLCES 500000U
+#define ONE_MIL_CYLCES 1000000U
+#define HALF_A_SECOND HALF_MIL_CYLCES * osClockRate / ONE_MIL_CYLCES
+
+OSPifRam __osContPifRam;
+u8 __osContLastCmd;
+u8 __osMaxControllers;
+
+OSTimer __osEepromTimer;
+OSMesgQueue __osEepromTimerQ ALIGNED(0x8);
+OSMesg __osEepromTimerMsg;
+
+s32 __osContinitialized = FALSE;
+extern OSTime osClockRate;
+
+s32 osContInit(OSMesgQueue* mq, u8* bitpattern, OSContStatus* data) {
+    OSMesg dummy;
+    s32 ret = 0;
+    OSTime t;
+    OSTimer mytimer;
+    OSMesgQueue timerMesgQueue;
+
+    if (__osContinitialized) {
+        return 0;
+    }
+
+    __osContinitialized = TRUE;
+
+    t = osGetTime();
+    if (HALF_A_SECOND > t) {
+        osCreateMesgQueue(&timerMesgQueue, &dummy, 1);
+        osSetTimer(&mytimer, HALF_A_SECOND - t, 0, &timerMesgQueue, &dummy);
+        osRecvMesg(&timerMesgQueue, &dummy, OS_MESG_BLOCK);
+    }
+
+    __osMaxControllers = 4;
```

Comment:
MAXCONTROLLERS should still be a valid macro in base libultra, not that the number 4 isn't unintuitive